File: implementations/typescript/src/convert.ts
import { AstNode, BlockNode, DocumentNode } from "./ast.js";

type V = string | number | boolean | V[] | { [k: string]: V };

function nodeToValue(node: AstNode): V {
  if (node.kind === "string") return node.value;
  if (node.kind === "number") return node.value;
  if (node.kind === "boolean") return node.value;
  if (node.kind === "array") return node.elements.map(nodeToValue);
  throw new Error(`unsupported node kind ${node.kind}`);
}

function uniqChildren(block: BlockNode): BlockNode[] {
  const seen = new Set<BlockNode>();
  const out: BlockNode[] = [];
  for (const child of Object.values(block.blocks)) {
    if (seen.has(child)) continue;
    seen.add(child);
    out.push(child);
  }
  for (const child of block.named_blocks ?? []) {
    if (seen.has(child)) continue;
    seen.add(child);
    out.push(child);
  }
  return out;
}

function blockToObject(block: BlockNode): { [k: string]: V } {
  const result: { [k: string]: V } = {};
  for (const [k, v] of Object.entries(block.properties)) insertPath(result, k, nodeToValue(v));

  const children = uniqChildren(block);
  for (const child of children.filter((c) => c.argument === undefined)) {
    const existing = result[child.name];
    const childObj = blockToObject(child);
    if (existing && typeof existing === "object" && !Array.isArray(existing)) {
      result[child.name] = { ...childObj, ...(existing as { [k: string]: V }) };
    } else result[child.name] = childObj;
  }

  for (const child of children.filter((c) => c.argument !== undefined)) {
    const base = namedBase(child.name);
    const arg = child.argument as string;
    const branch = result[base];
    const parent = branch && typeof branch === "object" && !Array.isArray(branch) ? (branch as { [k: string]: V }) : {};
    parent[arg] = blockToObject(child);
    result[base] = parent;
  }
  return result;
}

function insertPath(target: { [k: string]: V }, key: string, value: V): void {
  const parts = key.split(".");
  if (parts.length === 1) {
    if (Object.prototype.hasOwnProperty.call(target, key)) throw new Error(`duplicate key '${key}'`);
    target[key] = value;
    return;
  }
  const head = parts[0]!;
  const current = target[head];
  if (current !== undefined && (typeof current !== "object" || Array.isArray(current))) throw new Error(`key conflict at '${head}'`);
  const branch = (current as { [k: string]: V } | undefined) ?? {};
  insertPath(branch, parts.slice(1).join("."), value);
  target[head] = branch;
}

export function toObject(document: DocumentNode): { [k: string]: V } {
  const out: { [k: string]: V } = {};
  for (const block of document.blocks) {
    if (block.argument !== undefined) out[namedBase(block.name)] = { [block.argument]: blockToObject(block) };
    else out[block.name] = blockToObject(block);
  }
  return out;
}

function namedBase(name: string): string { return name === "region" ? "regions" : name; }

function esc(s: string): string {
  return s.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n").replace(/\t/g, "\\t");
}

function emitYAML(v: V, indent = 0): string {
  const pad = "  ".repeat(indent);
  if (Array.isArray(v)) {
    return v.map((item) => (typeof item === "object" ? `${pad}-\n${emitYAML(item as V, indent + 1)}` : `${pad}- ${emitScalar(item)}`)).join("\n");
  }
  if (typeof v === "object") {
    return Object.keys(v).sort().map((k) => {
      const item = (v as { [k: string]: V })[k];
      if (typeof item === "object") return `${pad}${k}:\n${emitYAML(item as V, indent + 1)}`;
      return `${pad}${k}: ${emitScalar(item)}`;
    }).join("\n");
  }
  return `${pad}${emitScalar(v)}`;
}

function emitTOML(v: { [k: string]: V }): string {
  const out: string[] = [];
  const walk = (obj: { [k: string]: V }, prefix?: string): void => {
    for (const k of Object.keys(obj).sort()) {
      const item = obj[k];
      if (typeof item === "object" && !Array.isArray(item)) continue;
      out.push(`${k} = ${emitScalar(item as V)}`);
    }
    for (const k of Object.keys(obj).sort()) {
      const item = obj[k];
      if (!(typeof item === "object") || Array.isArray(item)) continue;
      const sec = prefix ? `${prefix}.${k}` : k;
      if (out.length > 0) out.push("");
      out.push(`[${sec}]`);
      walk(item as { [k: string]: V }, sec);
    }
  };
  walk(v);
  return out.join("\n");
}

function emitHCL(v: V, indent = 0): string {
  const pad = "  ".repeat(indent);
  if (typeof v !== "object" || Array.isArray(v)) return `${pad}${emitScalar(v as V)}`;
  const lines: string[] = [];
  for (const k of Object.keys(v).sort()) {
    const item = (v as { [k: string]: V })[k];
    if (typeof item === "object" && !Array.isArray(item)) {
      lines.push(`${pad}${k} {`);
      lines.push(emitHCL(item, indent + 1));
      lines.push(`${pad}}`);
    } else lines.push(`${pad}${k} = ${emitScalar(item)}`);
  }
  return lines.join("\n");
}

function emitScalar(v: V): string {
  if (typeof v === "string") return `"${esc(v)}"`;
  if (typeof v === "number") return `${v}`;
  if (typeof v === "boolean") return v ? "true" : "false";
  if (Array.isArray(v)) return `[${v.map(emitScalar).join(", ")}]`;
  return "{}";
}

export function toYAML(document: DocumentNode): string { return emitYAML(toObject(document)); }
export function toTOML(document: DocumentNode): string { return emitTOML(toObject(document)); }
export function toHCL(document: DocumentNode): string { return emitHCL(toObject(document)); }

File: implementations/rust/src/convert.rs
use std::collections::{BTreeMap, BTreeSet};

use serde::Serialize;

use crate::ast::{AstNode, BlockNode, DocumentNode};

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(untagged)]
pub enum Value { S(String), N(f64), B(bool), A(Vec<Value>), O(BTreeMap<String, Value>) }

pub fn to_object(doc: &DocumentNode) -> BTreeMap<String, Value> {
    if let Some(root) = &doc.root_value {
        let mut wrapped = BTreeMap::new();
        wrapped.insert("root".to_string(), node_to_value(root));
        return wrapped;
    }
    let mut out = BTreeMap::new();
    for b in &doc.blocks {
        if let Some(arg) = &b.argument {
            let mut arg_map = BTreeMap::new();
            arg_map.insert(arg.clone(), Value::O(block_to_map(b)));
            out.insert(named_base(&b.name), Value::O(arg_map));
        } else {
            out.insert(b.name.clone(), Value::O(block_to_map(b)));
        }
    }
    out
}

pub fn to_yaml(doc: &DocumentNode) -> String {
    match serde_yaml::to_string(&Value::O(to_object(doc))) {
        Ok(s) => s.trim_start_matches("---\n").trim_end().to_string(),
        Err(_) => String::new(),
    }
}

pub fn to_toml(doc: &DocumentNode) -> String {
    toml::to_string(&to_object(doc)).unwrap_or_default().trim_end().to_string()
}

pub fn to_hcl(doc: &DocumentNode) -> String {
    emit_hcl(&Value::O(to_object(doc)), 0)
}

fn block_to_map(block: &BlockNode) -> BTreeMap<String, Value> {
    let mut out = BTreeMap::new();
    for (k, v) in &block.properties { insert_path(&mut out, k, node_to_value(v)); }

    let children = uniq_children(block);
    for c in children.iter().filter(|x| x.argument.is_none()) {
        let mut child = block_to_map(c);
        if let Some(Value::O(existing)) = out.get(&c.name) {
            for (k, v) in existing { child.insert(k.clone(), v.clone()); }
        }
        out.insert(c.name.clone(), Value::O(child));
    }
    for c in children.iter().filter(|x| x.argument.is_some()) {
        let arg = c.argument.clone().unwrap_or_default();
        let base = named_base(&c.name);
        let mut parent = match out.get(&base) { Some(Value::O(v)) => v.clone(), _ => BTreeMap::new() };
        parent.insert(arg, Value::O(block_to_map(c)));
        out.insert(base, Value::O(parent));
    }
    out
}

fn uniq_children(block: &BlockNode) -> Vec<BlockNode> {
    let mut out = Vec::new();
    let mut seen = BTreeSet::new();
    for c in block.blocks.values().chain(block.named_blocks.iter()) {
        let id = if let Some(a) = &c.argument { format!("{}:{}", c.name, a) } else { c.name.clone() };
        if seen.insert(id) { out.push(c.clone()); }
    }
    out
}

fn node_to_value(node: &AstNode) -> Value {
    match node {
        AstNode::String { value, .. } => Value::S(value.clone()),
        AstNode::Number { value, .. } => Value::N(*value),
        AstNode::Boolean { value, .. } => Value::B(*value),
        AstNode::Array { elements, .. } => Value::A(elements.iter().map(node_to_value).collect()),
        AstNode::Block { block, .. } => Value::O(block_to_map(block)),
    }
}

fn emit_hcl(v: &Value, indent: usize) -> String {
    let pad = "  ".repeat(indent);
    let Value::O(obj) = v else { return format!("{}{}", pad, scalar(v)) };
    obj.iter().map(|(k, item)| {
        if let Value::O(_) = item { format!("{}{} {{\n{}\n{}}}", pad, k, emit_hcl(item, indent + 1), pad) }
        else { format!("{}{} = {}", pad, k, scalar(item)) }
    }).collect::<Vec<_>>().join("\n")
}

fn scalar(v: &Value) -> String {
    match v {
        Value::S(s) => format!("\"{}\"", s.replace('\\', "\\\\").replace('"', "\\\"").replace('\n', "\\n").replace('\t', "\\t")),
        Value::N(n) => n.to_string(),
        Value::B(b) => if *b { "true".into() } else { "false".into() },
        Value::A(a) => format!("[{}]", a.iter().map(scalar).collect::<Vec<_>>().join(", ")),
        Value::O(_) => "{}".into(),
    }
}

fn insert_path(target: &mut BTreeMap<String, Value>, key: &str, value: Value) {
    let parts: Vec<&str> = key.split('.').collect();
    if parts.len() == 1 {
        if target.contains_key(key) { panic!("duplicate key"); }
        target.insert(key.to_string(), value);
        return;
    }
    let head = parts[0].to_string();
    let branch = match target.get(&head) {
        Some(Value::O(v)) => v.clone(),
        Some(_) => panic!("key conflict"),
        None => BTreeMap::new(),
    };
    let mut next = branch;
    insert_path(&mut next, &parts[1..].join("."), value);
    target.insert(head, Value::O(next));
}

fn named_base(name: &str) -> String {
    if name.ends_with('s') { name.to_string() } else { format!("{name}s") }
}
